pytorch_tabnet/callbacks.py

```diff
@@ -0,0 +1,223 @@
+        pass
+
+    def on_train_begin(self, logs=None):
+        pass
+
+    def on_train_end(self, logs=None):
+        pass
+
+
+@dataclass
+class CallbackContainer:
+    """
+    Container holding a list of callbacks.
+    """
+
+    callbacks: List[Callback] = field(default_factory=list)
+
+    def append(self, callback):
+        self.callbacks.append(callback)
+
+    def set_params(self, params):
+        for callback in self.callbacks:
+            callback.set_params(params)
+
+    def set_trainer(self, trainer):
+        self.trainer = trainer
+        for callback in self.callbacks:
+            callback.set_trainer(trainer)
+
+    def on_epoch_begin(self, epoch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_epoch_begin(epoch, logs)
+
+    def on_epoch_end(self, epoch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_epoch_end(epoch, logs)
+
+    def on_batch_begin(self, batch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_batch_begin(batch, logs)
+
+    def on_batch_end(self, batch, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_batch_end(batch, logs)
+
+    def on_train_begin(self, logs=None):
+        logs = logs or {}
+        logs["start_time"] = time.time()
+        for callback in self.callbacks:
+            callback.on_train_begin(logs)
+
+    def on_train_end(self, logs=None):
+        logs = logs or {}
+        for callback in self.callbacks:
+            callback.on_train_end(logs)
+
+
+@dataclass
+class EarlyStopping(Callback):
+    """EarlyStopping callback to exit the training loop if early_stopping_metric
+    does not improve by a certain amount for a certain
+    number of epochs.
+
+    Parameters
+    ---------
+    early_stopping_metric : str
+        Early stopping metric name
+    is_maximize : bool
+        Whether to maximize or not early_stopping_metric
+    tol : float
+        minimum change in monitored value to qualify as improvement.
+        This number should be positive.
+    patience : integer
+        number of epochs to wait for improvment before terminating.
+        the counter be reset after each improvment
+
+    """
+
+    early_stopping_metric: str
+    is_maximize: bool
+    tol: float = 0.0
+    patience: int = 5
+
+    def __post_init__(self):
+        self.best_epoch = 0
+        self.stopped_epoch = 0
+        self.wait = 0
+        self.best_weights = None
+        self.best_loss = np.inf
+        if self.is_maximize:
+            self.best_loss = -self.best_loss
+        super().__init__()
+
+    def on_epoch_end(self, epoch, logs=None):
+        current_loss = logs.get(self.early_stopping_metric)
+        if current_loss is None:
+            return
+        if (self.is_maximize and (current_loss - self.best_loss) > self.tol) or (
```

Comment:
Maybe this can be made clearer with something like this:
```suggestion
        loss_change = current_loss - self.best_loss
        max_improved = (self.is_maximize and loss_change > self.tol)
        min_improved = (self.is_maximize and loss_change < self.tol)
        if  max_improved or min_improved:
```

Reply:
done